File: DataStructures/cpp/single_linked_list.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
};

class SingleLinkedList {
private:
    Node* head;

public:
    SingleLinkedList(int data) {
        if(data){
            head = new Node{data, nullptr};
        }
        else {
            cout << "initial data missing";
        }
    }

    ~SingleLinkedList() {
        Node* current = head;
        while (current != nullptr) {
            Node* next = current->next;
            delete current;
            current = next;
        }
    }

    void print() {
        Node* current = head;
        while (current != nullptr) {
            cout << current->data << " ";
            current = current->next;
        }
        cout << endl;
    }

    void addAtTop(int data) {
        Node* newNode = new Node{data, head};
        head = newNode;
    }

    void append(int data){
        Node *cur = head;
        while (cur->next!=nullptr){
            cur=cur->next;
        }
        Node* newNode = new Node{data, nullptr};
        cur->next = newNode;
    }

    void addInBTW(int p1, int data, int p2) {
        if (head == nullptr) {
            return;
        }
        Node* cur = head;
        while (cur != nullptr && cur->next != nullptr) {
            if (cur->data == p1 && cur->next->data == p2) {
                Node* newNode = new Node{data, cur->next};
                cur->next = newNode;
                // cout << "Insertion Done in BTW\n";
                return;
            }
            cur = cur->next;
        }
        cout << "Could not find " << p1 << " and " << p2 << " in sequence." << endl;
    }

    void insert(int index,int data){
        if(index==0){
            addAtTop(data);
            return;
        }
        if (index==-1){
            append(data);
        }
        Node *cur = head;
        for(int i=0; i<index-1;i++){
            cur=cur->next;
        }
        if(cur==nullptr){
            cout << "List out of index";
            return;
        }
        Node *newnode = new Node{data,cur->next};
        cur->next = newnode;
    }

    void deleteAtTop(){
        Node *newnode = head->next;
        delete head;
        head = newnode;
    }

    void pop(){
        Node *cur = head;
        while(cur->next->next!=nullptr){
            cur = cur->next;
        }
        delete cur->next;
        cur->next = nullptr;

    }

    void pop(int index){
        if(index==0){
            deleteAtTop();
            return;
        }

        if(index==-1){
            pop();
            return;
        }

        Node *cur = head;

        for(int i=0;cur!=nullptr && i<index-1;i++){
            cur = cur->next;
        }

        if (cur == nullptr || cur->next == nullptr) {
                return;
            }

        Node *tmp = cur->next;
        cur->next = tmp->next;
        delete tmp;
        return;
    }

    void remove(int data){
        if (head->data == data) {
            deleteAtTop();
            return;
        }
        Node *cur = head;
        while(cur->next!=nullptr && cur->next!=nullptr){
            if(cur->next->data == data){
                Node *tmp = cur->next;
                cur->next = tmp->next;
                delete tmp;
            }
            else{
            cur = cur->next;
        }}
    }
};

int main() {
    SingleLinkedList list(2);
    list.print();

    list.addAtTop(1);
    list.print();

    list.append(3);
    list.print();

    list.insert(3, 5);
    list.print();

    list.addInBTW(3, 4, 5);
    list.print();

    list.deleteAtTop();
    list.print();

    list.pop();
    list.print();

    list.pop(2);
    list.print();

    list.remove(3);
    list.print();

    return 0;
}

File: DataStructures/Java/SingleLinkedList.java
// Single Linked List implementation in Java (java 21.0.5)
// java 21.0.5 2024-10-15 LTS
// Java(TM) SE Runtime Environment (build 21.0.5+9-LTS-239)
// Java HotSpot(TM) 64-Bit Server VM (build 21.0.5+9-LTS-239, mixed mode, sharing)

package Java;

interface LinkedList {
    void addAtTop(int data); // Insert data at beginning
    void addAtEnd(int data); // Insert data at End
    void addAtPos(int data, int pos); // Insert data at Position
    void deleteAtTop(); // Delete data at Top
    void deleteAtEnd(); // Delete data at End
    void deleteAtPos(int pos); // Delete data at Position
    void display(); //  Display the list
}

// Node: The SINGLE~LINKED~LIST
class Node {

    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class SingleLinkedListImpl implements LinkedList {

    // Single Linked List Implementation
    // Usage: SingleLinkedListImpl list = new SingleLinkedListImpl();

    // The First Node
    private Node head;

    // Insert data at top
    // Usage: list.addAtTop(10)
    public void addAtTop(int data) {
        Node newNode = new Node(data);
        newNode.next = head;
        head = newNode;
    }

    // Insert data at end
    // list.addAtEnd(20);
    public void addAtEnd(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node temp = head;
        while (temp.next != null) temp = temp.next;
        temp.next = newNode;
    }

    // Insert data at Pos
    // list.addAtPos(15, 2);
    public void addAtPos(int data, int pos) {
        if (pos < 0) return;
        if (pos == 0) {
            addAtTop(data);
            return;
        }

        Node newNode = new Node(data);
        Node temp = head;
        for (int i = 0; i < pos - 1 && temp != null; i++) {
            temp = temp.next;
        }
        if (temp == null) return;
        newNode.next = temp.next;
        temp.next = newNode;
    }

    // Delete data at Top
    // list.deleteAtTop();
    public void deleteAtTop() {
        if (head != null) head = head.next;
    }

    // Delete data at end
    // list.deleteAtEnd()
    public void deleteAtEnd() {
        if (head == null) return;
        if (head.next == null) {
            head = null;
            return;
        }
        Node temp = head;
        while (temp.next.next != null) {
            temp = temp.next;
        }
        temp.next = null;
    }

    // Delete data at Position
    // list.deleteAtPos(1);
    public void deleteAtPos(int pos) {
        if (pos < 0 || head == null) return;
        if (pos == 0) {
            deleteAtTop();
            return;
        }
        Node temp = head;
        for (int i = 0; i < pos - 1 && temp.next != null; i++) {
            temp = temp.next;
        }
        if (temp.next == null) return;
        temp.next = temp.next.next;
    }

    // Display the list
    // list.display();
    public void display() {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " -> ");
            temp = temp.next;
        }
        System.out.println("null");
    }
}

// Main class
public class SingleLinkedList {

    public static void main(String[] args) {
        System.out.println("Linked List Demo:");
        SingleLinkedListImpl list = new SingleLinkedListImpl();

        list.addAtTop(10);
        list.addAtEnd(20);
        list.addAtEnd(30);
        list.addAtPos(15, 1);
        list.display();

        list.deleteAtTop();
        list.deleteAtEnd();
        list.deleteAtPos(1);
        list.display();
    }
}
